src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -3523,19 +3544,136 @@ WorkerSortClauseList(MultiExtendedOp *originalOpNode,
+ *
+ * Note the function modifies workerExtendedOpNodes targetList, contents
+ * of SortGroupClause items in sortClauseList, and nextSortGroupRefIndex.
+ */
+static void
+AdjustWorkerSortClausesAndTargetList(MultiExtendedOp *workerExtendedOpNode,
+									 MultiExtendedOp *originalOpNode,
+									 List *sortClauseList,
+									 Index *nextSortGroupRefIndex)
+{
+	ListCell *sortClauseCell = NULL;
+	List *workerTargetList = workerExtendedOpNode->targetList;
+	List *originalTargetList = originalOpNode->targetList;
+	HASHCTL hashCtl;
+	HTAB *targetEntryMap = NULL;
+
+	MemSet(&hashCtl, 0, sizeof(hashCtl));
+	hashCtl.keysize = sizeof(Index);
+	hashCtl.entrysize = sizeof(Index);
+
+	hashCtl.hcxt = CurrentMemoryContext;
+
+	targetEntryMap = hash_create("target entry sort group map", 8,
+								 &hashCtl,
+								 HASH_ELEM | HASH_BLOBS);
+
+	Assert(sortClauseList != NIL);
+
+	foreach(sortClauseCell, sortClauseList)
+	{
+		SortGroupClause *sgClause = (SortGroupClause *) lfirst(sortClauseCell);
+		TargetEntry *targetEntry = get_sortgroupclause_tle(sgClause, originalTargetList);
+		bool containsAggregate = contain_agg_clause((Node *) targetEntry->expr);
+		Index oldIndex = sgClause->tleSortGroupRef;
+		Index *newIndexPtr = NULL;
+		bool found = false;
+		bool useOriginalTargetEntry = false;
+		Expr *targetExpr = targetEntry->expr;
+
+
+		/* we are only interested in target entries containing aggregates */
+		if (!containsAggregate)
+		{
+			continue;
+		}
+
+		/*
+		 * If the target expression is not an Aggref, it is either an expression
+		 * on a single aggregate, or expression containing multiple aggregates.
+		 * Worker query mutates these target entries to have a naked target entry
+		 * per aggregate function. We want to use original target entries if this
+		 * the case.
+		 * If the original target expression is an avg aggref, we also want to use
+		 * original target entry.
+		 */
+		if (!IsA(targetExpr, Aggref))
+		{
+			useOriginalTargetEntry = true;
+		}
+		else
+		{
+			Aggref *aggNode = (Aggref *) targetExpr;
+			AggregateType aggregateType = GetAggregateType(aggNode->aggfnoid);
+			if (aggregateType == AGGREGATE_AVERAGE)
+			{
+				useOriginalTargetEntry = true;
+			}
+		}
+
+		if (!useOriginalTargetEntry)
+		{
+			continue;
+		}
+
+		/* here we only have target entries that we want to use in the original form */
+		newIndexPtr = (Index *) hash_search(targetEntryMap, &oldIndex, HASH_ENTER,
+											&found);
+
+		/*
+		 * if we already have new target entry created for this sort group reference,
+		 * just use that reference, otherwise create a new target entry in worker
+		 * target list.
+		 */
+		if (found)
+		{
+			sgClause->tleSortGroupRef = (*newIndexPtr);
+		}
+		else
+		{
+			TargetEntry *newTargetEntry = copyObject(targetEntry);
+			newTargetEntry->ressortgroupref = *nextSortGroupRefIndex;
+			sgClause->tleSortGroupRef = *nextSortGroupRefIndex;
+			newTargetEntry->resjunk = false;
```

Comment:
Hmm, I was thinking that we'd use `resjunk = true`. Do we actually need to pull that column from the worker? Aren't we already using the correct target entries on the coordinator?

Reply:
I was thinking that too, but master query did not play nice with it.  It expected to see this column in worker query results.  Perhaps we should fix this altogether.

Reply:
It turned out to be easy fix in multi_master_planner